035-cluster-wide-volume-usage-quota-managemnet.md

```diff
@@ -0,0 +1,129 @@
+clients. This is problematic with respect
+to [KIP-112](https://cwiki.apache.org/confluence/display/KAFKA/KIP-112%3A+Handle+disk+failure+for+JBOD) as the broker
+will take partitions offline when the volume they are stored on runs out of space. Which in the case of unbalanced usage
+between volumes can lead to a volume running out of storage without throttling being applied. The broker will go-offline
+if all volumes are unavailable.
+
+Configuring the soft and hard limit thresholds is currently done in terms of bytes consumed, which forces users
+deploying the plugin to calculate a threshold for throttling and would require separate thresholds per volume to avoid
+out of space issues.
+
+## Proposal
+
+### Leverage kafka to distribute volume usage metrics throughout the cluster.
+
+By publishing per volume usage metrics to a compacted topic keyed by the broker ID each instance of the plugin will be able
+to start and quickly determine the state of the cluster and consistently apply throttling regardless of where the client
+is connected.
+
+[KIP-257](https://cwiki.apache.org/confluence/display/KAFKA/KIP-257+-+Configurable+Quota+Management) Defines the quota
+API and defines quotas in terms of how much of a delay to apply to a given produce request. This leads to defining the
+possible states of the plugin as:
+`OPEN`, `THROTTLE` and `PAUSE` to accurately reflect the effects of transitioning between states.
+
+#### Limit Types
+
+1. Consumed space limit: Triggers if the consumed space of a volume breaches the value. Candidate for deprecation.
+2. Minimum Free Bytes: Triggers if the amount of free space on a volume drops below the configured level. Particularly
+   useful for hard limits as an absolute minimum of free space.
+3. Minimum Free Percentage. Triggers if the amount of free space on the volume drops below the configured proportion of
+   the Volumes total capacity.
+
+#### The quota plugin data publisher should:
+
+- Collect local volume usage
+- Publish usage to an internal compacted topic keyed by broker ID.
+- Collect and publish usage on start up and periodically after that.
+- Connect as an internal service using TLS and client certificates to connect to the Control Plane listener.
+- Publish metrics showing its connection status and a counter of the number of publication errors - to allow the
+  detection of connectivity issues.
+- Publish Volume Information with the following [schema](#Message-schema)
+
+#### The quota plugin consumer should:
+
+- Should apply the largest delay caused by each disk which breaches the soft limit.
+    - The amount of throttling should be proportional to how much of the difference between the soft and hard limit
+      remains for a given Volume. Applying more throttle the closer to the hard limit things are.
+- Should apply a `PAUSE` of publication if *any* Volume in the cluster breaches the hard limit
```

Comment:
I've clarified the wording. Yes it should block external producers rather than stop publishing new metrics. 

I think the storage requirement of the metrics is roughly steady state. Given the number of messages published is 1 per broker and the number of volumes per broker is static in 99% of cases.